File: module-04-waiting/js/implicity_vs_explicity.js
const { chromium } = require('playwright');

(async () => {
  const browser = await chromium.launch();
  const page    = await browser.newPage();
  await page.goto("https://example.com/delay-element");

  // Playwright 沒隱式等待，但可模擬
  page.setDefaultTimeout(5000);

  // 測量「自動等待」的時間 (第一次 selector 錯誤時會重試)
  let t0 = Date.now();
  await page.$("#delayed");
  let t1 = Date.now();
  console.log(`Auto-wait took ${(t1-t0)/1000}s`);

  // 測量顯式等待
  await page.goto("https://example.com/delay-element"); // reload
  let t2 = Date.now();
  await page.waitForSelector("#delayed", { timeout: 5000 });
  let t3 = Date.now();
  console.log(`Explicit wait took ${(t3-t2)/1000}s`);

  await browser.close();
})();

File: module-04-waiting/py/implicity_vs_explicity.py
import time
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

driver = webdriver.Chrome(
    service=Service(ChromeDriverManager().install())
)
driver.implicitly_wait(5)  # 隱式等待 5 秒
driver.get("https://example.com/delay-element")

# 測量隱式等待搜尋時間
t0 = time.time()
driver.find_element(By.ID, "delayed")  # 若元素未現，隱式等待
t1 = time.time()
print(f"Implicit wait took {t1-t0:.2f} seconds")

# 測量顯式等待搜尋時間
driver.refresh()  # 重載頁面
t2 = time.time()
WebDriverWait(driver, 5).until(EC.presence_of_element_located((By.ID, "delayed")))
t3 = time.time()
print(f"Explicit wait took {t3-t2:.2f} seconds")

driver.quit()
